doc: Re-order and edit RTIO page; add glossary

mmphys1 opened the pull request:
<!--

Thank you for submitting a PR to ARTIQ!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

You can also read more about contributing to ARTIQ in this document:
https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#contributing-code

Based on https://raw.githubusercontent.com/PyCQA/pylint/master/.github/PULL_REQUEST_TEMPLATE.md
-->

# ARTIQ Pull Request

## Description of Changes

- Changed the ordering of technical topics in ```rtio.rst``` ; 
- Added a glossary of terms for RTIO.

The motivation for this PR is to make it easier for newcomers to understand the RTIO principles necessary for using Artiq.

Currently, the ```rtio.rst``` page is indeed useful and thorough. However, when I put myself in the shoes of a first-user of Artiq, I noticed some concepts were introduced before they were necessary. For example, the timeline cursor and associated technical details are introduced before the timeline section. Therefore, to support users new to Artiq, I implemented the changes as bulleted above.
 
To keep the writing consistent did necessitate significant editing, although I strive not to have eliminated any content already there—only rewording. I have taken care to ensure no regressions are introduced; naturally, this means I built the docs by running ```make html``` accordingly. (I'm using Nix, although I didn't use the specific ```nix build``` script below).

*Context: when I was using Artiq for the first time myself a while back, I wrote notes on what I did and didn’t understand, following which I did this re-ordering for myself. This PR aims to take part of what I learnt and update the docs accordingly. To keep contributions atomic, I’ve just focused on the RTIO page.*

### Related Issue

Expands upon #2451.
<!-- 
If this PR fixes a particular issue, use the following to automatically close that issue
once this PR gets merged:

Closes #XXX 
-->

## Type of Changes

<!-- Leave ONLY the corresponding lines for the applicable type of change: -->
|   | Type |
| ------------- | ------------- |
| ✓  | :scroll: Docs |

## Steps (Choose relevant, delete irrelevant before submitting)

### All Pull Requests

- [x] Use correct spelling and grammar.
- [x] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`nix build .#artiq-manual-html; nix build .#artiq-manual-pdf`) to ensure no errors.

### Git Logistics

- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

architeuthidae commented:
Hmm, there's a lot more changed here than just ordering.

A glossary may be a good idea, if the existing in-line definitions are difficult to keep an overview of. Is this a problem you see in other sections as well? Is it only the RTIO page where you'd like to see a glossary section, or might it in your opinion even be useful to have an overarching glossary _page_ in the addenda?  

> For example, the timeline cursor and associated technical details are introduced before the timeline section.

Can you clarify what you mean by this? The "Timeline and terminology" section is the first section in the current page, and your version also provides the glossary _before_ giving an illustration of what's meant by the timeline.  

Broadly I disagree with the stylistic choice of a shift from a reference text ("the state of things is such, such, and such; to accomplish this, one must do such and such") to a more conversational tutorial ("now let's think about this, now let's try this, how can we do this?"). A more step-by-step tutorial-like tone is taken in the following "getting started" pages; this page, to me, is meant to serve as the primary reference text for ARTIQ abstractions, and is more efficiently presented that way. This approach adds a lot of fluff and extraneous phrasing; I'm not convinced it is strictly less confusing, and even if it contributes to readability at the very first readthrough, having to sort through fluff reduces utility every following time the text is referenced. 

Some of the additions to the first section also seem out of place; this doesn't necessarily seem like the place to define "core device", for example.

mmphys1 commented:
Thanks for the feedback, much appreciated.

> Hmm, there's a lot more changed here than just ordering.

Yes, I do emphasise that I wasn't trying to sneak something in---as I said, I wanted to keep the writing consistent after re-ordering. All changes serve the re-ordering and motivation: perhaps I could suggest an addition to the change description to reflect that? :

- Technical writing to fit the re-ordering of information.

Naturally, I'm happy to rename the PR accordingly.

> Is [the lack of a glossary] a problem you see in other sections as well? Is it only the RTIO page where you'd like to see a glossary section, or might it in your opinion even be useful  to have an overarching glossary *page* in the addenda?

Yes, a separate glossary page may be useful. Although as a first step, I added a glossary for RTIO only since:

- The majority of new technical terms are introduced there;
- it's likely the first introduction to the Artiq documentation and many concepts, [as stated in the FAQ](https://m-labs.hk/artiq/manual-beta/faq.html#use-this-documentation). (Andecotally) most users start on already working Artiq setups.

Adding a separate glossary page might be quite time-consuming, since one probably needs to read a large proportion of the docs in one go before doing so, to keep the glossary consistent. 

> For example, the timeline cursor and associated technical details are introduced before the timeline section.

> > Can you clarify what you mean by this? The "Timeline and  terminology" section is the first section in the current page, and your  version also provides the glossary *before* giving an illustration of what's meant by the timeline.

Poorly clarified by me, sorry. I meant that several new technical notes and details are introduced without being used before the timeline example and diagram. I should've therefore written 'without being used in the following example/discussions', rather than 'before the timeline section'. As a reference, that's of course fine to give many technical details. As a tutorial, it reflects less how new users build up a picture of how things are working, which I shall refer to in my next point. (Hence, regarding that I gave the glossary before the illustration, then it's a misunderstanding based on what I wrote before).

> Broadly I disagree with the stylistic choice of a shift  from a reference text [...] to a more conversational  tutorial [...].

I do agree with the need for a consistent tone. Although I did feel the incongruity when writing it, I went from the implicit [guidance in the current Artiq FAQ](https://m-labs.hk/artiq/manual-beta/faq.html#use-this-documentation). Namely, the RTIO is defined as a tutorial page, is under a tutorial heading in the docs sidebar, and is the first part of the manual that most new users will read. Further, from the perspective of the documentation writers serving the Artiq users, one may ask: is the purpose of the RTIO page as a tutorial or as a reference? It may be difficult to straddle both purposes simultaneously. There is the somewhat popular  https://diataxis.fr/, although I wouldn't take that as gospel. For these reasons, I went with tutorial-style discussion. Nevertheless, perhaps the tone itself can be made more congruous with the rest of the docs by making it more formal? In which case I'm happy to edit accordingly. Your opinions are appreciated there.

> Some of the additions to the first section also seem out  of place. This doesn't necessarily seem like the place to define "core  device", for example.

The core device is implicitly defined already in the current RTIO page:

>...a given experiment is composed of two types of “programs”: regular Python code that is executed on the host and ARTIQ *kernels* that are executed on a ***core device***.

As I said, I tried not to lose any information of what's already there in the docs. At another glance, I don't immediately spot what other additions are out of place---perhaps I ought to look with fresher eyes later.

I hope that what I've written is polite and kind, I do respect and appreciate the time and efforts of the Artiq developers and contributors. Thanks again.

architeuthidae commented:
> Adding a separate glossary page might be quite time-consuming, since one probably needs to read a large proportion of the docs in one go before doing so, to keep the glossary consistent.

Most of the current manual was written or re-written by me, so it's not all too big an ask, and it's an idea that hadn't occurred to me....  

> Namely, the RTIO is defined as a tutorial page, is under a tutorial heading in the docs sidebar, and is the first part of the manual that most new users will read. Further, from the perspective of the documentation writers serving the Artiq users, one may ask: is the purpose of the RTIO page as a tutorial or as a reference?

That's a fair point, it is categorized as a tutorial. I've considered moving it under the "ARTIQ Components" before, which does fit the material and presentation better, but it's somewhat of a compromise based on the fact that the manual is in reading order (ish) -- and unlike the core device or management system pages, I do somewhat think this one is necessary to read before the "Getting started" pages for best understanding. But by the diataxis system you link I'd categorize it as "explanation". 

A different question is whether it _should_ be. I'm not really convinced the tutorial format (coherence with FAQ aside, the FAQ can also be changed if you find it misleading!) makes sense, in particular because the material is not practical, it's conceptual; we are attempting to lay out the various abstractions ARTIQ uses, so that we can use those abstractions in discussion later. There are no actions here for the user to perform, only steps in understanding. 

Do you agree with the diataxis framework that explanation should come _after_ tutorial? I could see an argument for drastically shortening this page and combining it with the succeeding "Getting started" tutorial, and moving most of the explanation material to a later place in the manual. Alternatively, probably at least intention could be clarified (I'm thinking of an initial note to the effect of _skip to the next section if you want to dive in immediately, and come back here when you get confused about concepts or terminology._)

> I hope that what I've written is polite and kind, I do respect and appreciate the time and efforts of the Artiq developers and contributors. Thanks again.

Thank you very much too :) If I communicate bluntly here, it's for efficiency and clarity, never to reject discussion.

mmphys1 commented:
Got it. Let's put all these points together, and from there decide what to do next.

I agree and I think I've come to the same conclusion as you: drastically shorten this RTIO page. The RTIO *tutorial* can be made much shorter by splitting the reference parts and tutorial parts into separate pages. 

Suggested splitting:

1. 'What is ARTIQ Real-Time I/O?': Introduction and motivation to RTIO on separate page. This is reference style, but counts as part of tutorial. Brief mention of timeline on this page.

2. 'RTIO concepts': Timeline, terminology and the basic timeline example on a separate, tutorial-style page. This can be combined with the Getting started page.

   > Add a note: "*skip to the next section if you want to dive in immediately, and  come back here when you get confused about concepts or terminology.*"

   Yes, definitely agree to put that.

3. 'RTIO errors': Everything from 'Output errors and exceptions' in an explanation-style page (or could go under a reference heading in the current docs structure).

Win for everyone: tutorial is much shorter, and reference is more purposeful and succinct.

There are two catches:

1. new users might not read the reference pages fully until much later. The FAQ should be updated with according advice, to encourage them to read it early on when they start encountering ARTIQ gotchas.
2. The reference info for the RTIO concepts must find an appropriate place.

Therefore I actually neither agree nor disagree with diataxis framework that explanation must follow tutorial. As long as the appropriate (hyper)links direct the user, they shouldn't be lost in our particular use case.

-------

Addressing your other comments:

> I do somewhat think [the RTIO page] is necessary to read before the "Getting started" pages for best understanding.

100% agree, I think it's absolutely necessary.

> A different question is whether it *should* be [a tutorial].

Most users would like to progress towards the experiment they want to do, right? It motivates them to e.g. be able to make their laser pulse by appropriate driving of an optical modulator. I think it's not too much of a stretch to for most users to extend the pulse concepts in the timeline example to control their AOMs or EOMs or similar. Thus, tutorial for that timeline example makes sense to me. 

*Separate glossary page*: would be better in the long run and could be started quickly, since that's just a matter of finding tricky words in the manual via whatever quick methods (Ctrl+F, grep, AI...). Then it's a matter of how tolerable errors would be, for the arduous checks thereafter. 

*Location of glossary*: my own PR is sub-optimal here, because the glossary is half-way down the page and not exposed in the sidebar.  It's a compromise since it's probably not good to add a large change at once. Do you prefer to have a separate glossary page from now, or keep the current location of RTIO glossary as is?

-------

Can this PR be worked into a net improvement on the current docs? If yes, I can do the above (or leave some parts as-is if you think that's better); if no, then I welcome your suggestions, following which I can incorporate any text from this PR that may be useful for that.

If answer to last paragraph is yes, then tasks are:

- [ ] Split up pages.

- [ ] Edit pages to fit.

- [ ] Fit/edit glossary as desired.

mmphys1 commented:
Thanks for the feedback! I thought it would be most efficient to rewrite the PR following your feedback, adding in the glossary as well.

> I don't think splitting the page is necessary; the total  length presumably shouldn't change much. I'd suggest just rearranging  information into these sections *within* `rtio.rst`.

Indeed, the length didn't change much, even with adding in the glossary. I removed a lot, but I did add a little bit of repetition to make it fit (namely an extra timing diagram at the beginning, so as not to explain too much at once).

> if we can make [the glossary] work without tooltips, or relying on cross-linking  beyond what's already used in the manual, I'd prefer that.

I used the built-in glossary functionality of Sphinx, so it's cross-linking only what's in the manual, on the RTIO page. Only ARTIQ-specific stuff is in the glossary (maybe I should've wrote that in the commit...) since other such things could be googled (or similar).

> I agree with the general ordering of information here, but I still haven't seen a good reason to construct it as a tutorial. 

Hence, I wrote that in 'explanation' style. While I do believe tutorial style works better for beginners to ARTIQ, I reflected on my last reasonings and so would now agree that 'tutorial' style conflicts with the understanding of abstractions necessary for RTIO. More precisely, it may be that because of the abstract nature of RTIO concepts, for new users explanation is more effective to disseminate this knowledge than a tutorial, since one may require a lot of reflection to grasp those. Put simply, a tutorial is not so effective for something you can't easily practically demonstrate in reality. (If that makes sense).

Does the style fit better now? Having attempted myself, I can certainly appreciate the effort it takes to make good documentation, thanks for the feedback so far.

architeuthidae commented:
hmmm, I appreciate the work you've put into this so far, and this PR is definitely a structural improvement on what's there presently, but I have a lot of low-level stylistic nitpicks (sentence structure, grammar, clarity of phrasing, some content mistakes) which are inconvenient and difficult to try and all catch in a GitHub review. 

I wonder if you'd be open to closing this PR and letting me re-open a new version of it that links to this discussion? (I can promise that I'd do so a lot more promptly than I've responded to this one). That'd allow me to adopt this general structure and a lot of your particular order of information, incorporating the feedback you've given about the user experience, but handle things on the phrasing level myself. Then of course you'd be free to give feedback on that, point out misinterpretations of your contributions, etc. IMO it'd be the most efficient option here, and probably less work for both of us. What do you think? 

Also, an additional question: do you think the wiki-style :term: links will be very helpful to a new user? In the HTML they aren't visually distinguished from other types of links, which I dislike, it clutters up the page. My instinct is to prefer to keep the text clean -- people can ctrl+F if they need to, or just keep two windows open -- but I might be underestimating the utility they provide.

mmphys1 commented:
Ah, I'd checked before that maintainers could have write access, didn't realise that your account wasn't a maintainer, sorry!

Since I think everyone should have the opportunity to contribute, I'll make mostly suggestions rather than changes when reviewing on my branch, then we'll bring those changes back into upstream for others to review. Hope that works!

mmphys1 commented:
> (Commit history cleanup would probably be nice though.)

Cleaned. 

I messed up the rebase when I lost connection to my hotel wifi, so to fix it, I then:
- made a copy of the mm-docs-rtio branch (called mm-docs-rtio-clean); 
- cleaned up on that (rebase, made commits logical, sync/fetch from master, add authorship, reduce commit message length);
- overwrote the original mm-docs-rtio with mm-docs-rtio-clean. 

That is to say, while I've checked extensively that it should be clean now and merge with no problems and no commits lost (namely #2842 and #2820), I did **** up when getting there, so it may need a brief sanity check that nothing went awry...sorry for the added work, although it should be a few minutes extra only.

mmphys1 commented:
Fixed.

Following both git diff and diff of files in local editor on [my branch](https://github.com/mmphys1/artiq/tree/mm-docs-rtio) and [your branch](https://github.com/architeuthidae/artiq/tree/man-rtio-update) architeuthidae:man-rtio-update, both versions now agree.
